pkg/apiserver/internal/v1/offering.go

```diff
@@ -21,67 +21,133 @@ import (
 	if err := o.client.List(ctx, offeringList, listOptions...); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("listing offerings: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "listing offerings: %s", err.Error())
 	}
 
 	res, err = o.convertOfferingList(offeringList)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting OfferingList: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting OfferingList: %s", err.Error())
 	}
 	return
 }
 
 func (o offeringServer) Get(ctx context.Context, req *v1.OfferingGetRequest) (res *v1.Offering, err error) {
 	if err = o.validateGetRequest(req); err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	offering := &catalogv1alpha1.Offering{}
 	if err = o.client.Get(ctx, types.NamespacedName{
 		Name:      req.Name,
 		Namespace: req.Account,
 	}, offering); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("getting offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "getting Offering: %s", err.Error())
 	}
 	res, err = o.convertOffering(offering)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting Offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting Offering: %s", err.Error())
 	}
 	return
 }
 
+func (o offeringServer) Watch(req *v1.OfferingWatchRequest, stream v1.OfferingService_WatchServer) error {
+	listOptions, err := o.validateWatchRequest(req)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	watcher, err := o.dynamicClient.Resource(o.gvr).Namespace(req.Account).Watch(listOptions)
+	if err != nil {
+		return status.Errorf(codes.Internal, "watching offerings: %s", err.Error())
+	}
+	for {
+		select {
+		case <-stream.Context().Done():
+			return status.Error(codes.Internal, stream.Context().Err().Error())
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return status.Error(codes.Internal, "watch event channel was closed")
+			}
+			if req.OperationType != "" &&
+				req.OperationType != string(event.Type) {
+				continue
+			}
+			catalogOffering := &catalogv1alpha1.Offering{}
+			if err := o.scheme.Convert(event.Object, catalogOffering, nil); err != nil {
+				return status.Errorf(codes.Internal, "converting event.Object to Offering: %s", err.Error())
+			}
+			offering, err := o.convertOffering(catalogOffering)
+			if err != nil {
+				return status.Errorf(codes.Internal, "converting Offering: %s", err.Error())
+			}
+			any, err := ptypes.MarshalAny(offering)
+			if err != nil {
+				return status.Errorf(codes.Internal, "marshalling Offering to Any: %s", err.Error())
+			}
+			if err := stream.Send(&v1.Event{
+				Type:   string(event.Type),
+				Object: any,
+			}); err != nil {
+				return status.Errorf(codes.Internal, "sending Offering stream: %s", err.Error())
```

Comment:
here you should propagate the error code if the `err` satisfies the right interface, and use `codes.Internal` as a fallback.

```
type toGRPCStatus interface {
	GRPCStatus() *status.Status
}
```

Reply:
done